### ISSUE TITLE
cabal-install cmd line parser: silently accepts ambiguous prefixes

### ISSUE THREAD

**Issue by jwaldmann** [haskell/cabal#3089]
I was getting strange behaviour from
````
cabal install --with-hc-pkg=ghc-pkg-7.8.4 --with-hc=ghc-7.8.4
````
until I realized that one option is a prefix of the other, and was ignored. (The second option should be --with-compiler, I think)

I think this is wrong behaviour, instead cabal-install should complain loudly, and refuse to start.

This is for
```
cabal-install version 1.22.7.0
using version 1.22.4.0 of the Cabal library 
```

**Comment by 23Skidoo** [comment]
This is standard behaviour for GNU-style long options ([1](http://www.gnu.org/software/libc/manual/html_node/Argument-Syntax.html), [2](http://www.gnu.org/software/tar/manual/html_node/Long-Options.html), [3](https://docs.python.org/2/library/getopt.html)), and a lot of people rely on it.

**Comment by jwaldmann** [comment]
Really? 
```
tar --list --use-compress-program=bzip2 --use-compress=gzip something.tgz
tar: Conflicting compression options
```

**Comment by 23Skidoo** [comment]
OK, so you don't suggest removing abbreviations. Maybe we can do something.

**Comment by jwaldmann** [comment]
Yes, abbreviations are fine. I guess my "complaint" is: the options parser views
```
cabal --foo=bar1 --foo=bar2
```
as equivalent to
```
cabal --foo=bar1 
```
it silently ignores the value on the right, prefix-abbreviated or not. I think this is non-standard. It should complain and halt.

**Comment by 23Skidoo** [comment]
I'm not sure how non-standard this is, `gcc -O2 -O3` is the same as `gcc -O3`, for example. But for stuff like `--with-hc-pkg` it probably does make sense.

/cc @dcoutts
